github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/8395562?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/8395562?buildTab=artifacts#meta) on refs/heads/master @ [7953ed0f0058](https://github.com/cockroachdb/pebble/commits/7953ed0f0058):


```
=== PAUSE TestMeta/execution/random-007
=== CONT  TestMeta/execution/random-007
=== RUN   TestMeta/execution/random-009
=== PAUSE TestMeta/execution/random-009
=== CONT  TestMeta/execution/random-009
=== RUN   TestMeta/execution/standard-006
=== PAUSE TestMeta/execution/standard-006
=== CONT  TestMeta/execution/standard-006
=== RUN   TestMeta/execution/standard-011
=== PAUSE TestMeta/execution/standard-011
=== CONT  TestMeta/execution/standard-011
=== RUN   TestMeta/execution/random-002
=== PAUSE TestMeta/execution/random-002
=== CONT  TestMeta/execution/random-002
=== RUN   TestMeta/execution/standard-002
=== PAUSE TestMeta/execution/standard-002
=== CONT  TestMeta/execution/standard-002
=== RUN   TestMeta/execution/random-018
=== PAUSE TestMeta/execution/random-018
=== CONT  TestMeta/execution/random-018
=== RUN   TestMeta/execution/random-020
=== PAUSE TestMeta/execution/random-020
=== CONT  TestMeta/execution/random-020
=== RUN   TestMeta/execution/standard-016
=== PAUSE TestMeta/execution/standard-016
=== CONT  TestMeta/execution/standard-016
=== RUN   TestMeta/execution/random-006
=== PAUSE TestMeta/execution/random-006
=== CONT  TestMeta/execution/random-006
=== RUN   TestMeta/execution/random-024
=== PAUSE TestMeta/execution/random-024
=== CONT  TestMeta/execution/random-024
=== RUN   TestMeta/execution/standard-001
=== PAUSE TestMeta/execution/standard-001
=== CONT  TestMeta/execution/standard-001
=== RUN   TestMeta/execution/random-004
=== PAUSE TestMeta/execution/random-004
=== CONT  TestMeta/execution/random-004
=== RUN   TestMeta/execution/random-010
=== PAUSE TestMeta/execution/random-010
=== CONT  TestMeta/execution/random-010
=== RUN   TestMeta/execution/random-019
=== PAUSE TestMeta/execution/random-019
=== CONT  TestMeta/execution/random-019
=== RUN   TestMeta/execution/standard-015
=== PAUSE TestMeta/execution/standard-015
=== CONT  TestMeta/execution/standard-015
=== RUN   TestMeta/execution/standard-019
=== PAUSE TestMeta/execution/standard-019
=== CONT  TestMeta/execution/standard-019
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1674283162055892896 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/8455598?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/8455598?buildTab=artifacts#meta) on refs/heads/master @ [169e5dbe9992](https://github.com/cockroachdb/pebble/commits/169e5dbe9992):


```
        db.DeleteRange("lhheiuorjl@3", "loxqyf@3") // <nil> #2934
        db.DeleteRange("gyyvlrlpgfo@2", "qwgaiiz@2") // <nil> #2935
        db.DeleteRange("cabidxjkk@2", "rfithabgtz@8") // <nil> #2936
        db.DeleteRange("tnbdh@7", "ugavzitse@7") // <nil> #2937
        db.Set("cbmgihtgktv@2", "vjryjukorrzesxvajfb") // <nil> #2938
        // INFO: [JOB 74] WAL created 000049
        // INFO: [JOB 75] flushing 1 memtable to L0
        // INFO: [JOB 75] flushing: sstable created 000050
        // INFO: [JOB 75] flushed 1 memtable to L0 [000050] (3.5 K), in 0.0s (0.0s total), output rate 1.7 M/s
        // INFO: [JOB 75] WAL deleted 000044
        db.Flush() // <nil> #2939
        db.Delete("ltpfvrn@7") // <nil> #2940
        // INFO: [JOB 77] compacting(default) L0 [000043 000046 000050] (5.7 K) + L6 [000047] (3.9 K)
        db.SingleDelete("mrumchuebsn@5", false /* maybeReplaceDelete */) // <nil> #2941
        db.Merge("otxyxfscl@8", "xidiszyzjwbnxbn") // <nil> #2942
        db.Merge("bqceizrwhw@2", "oimjmsoggxdy") // <nil> #2943
        db.Merge("wfjjk", "zxnlbghp") // <nil> #2944
        batch19 = db.NewIndexedBatch() #2945
        db.Delete("vpeqb@1") // <nil> #2948
        batch19.DeleteRange("ewfvwtzs@8", "mmsjgqoko@8") // <nil> #2946
        batch19.Set("eyunjuuqoeb@8", "sihngnpto") // <nil> #2947
        db.Set("ystayojd@4", "xn") // <nil> #2950
        batch19.Delete("wddgxigoxj@6") // <nil> #2949
        // INFO: [JOB 77] compacting: sstable created 000051
        // INFO: [JOB 78] ingesting: sstable created 000052
        // INFO: [JOB 79] WAL created 000053
        // INFO: [JOB 80] flushing 1 memtable to L0
        // INFO: [JOB 80] flushing: sstable created 000054
        // INFO: [JOB 80] flushed 1 memtable to L0 [000054] (1.2 K), in 0.0s (0.0s total), output rate 1.9 M/s
        // INFO: [JOB 80] WAL deleted 000049
        // INFO: [JOB 77] compacted(default) L0 [000043 000046 000050] (5.7 K) + L6 [000047] (3.9 K) -> L6 [000051] (4.6 K), in 0.0s (0.0s total), output rate 564 K/s
        // INFO: [JOB 78] ingested L0:000052 (1.2 K)
        db.Ingest(batch19) // <nil> #2951
        db.Get("bdphbfmbne") // [""] pebble: not found #2952
        db.SingleDelete("hxthegwytyw@4", false /* maybeReplaceDelete */) // <nil> #2953
        db.Delete("zhmv@4") // <nil> #2954
        db.Delete("gzftkwxyi@8") // <nil> #2955
        db.Get("bmfprdgyll@8") // ["xptajlswudjdeznkkkl"] <nil> #2956
        db.Set("tylqxaf", "hloicpxmk") // <nil> #2957
        db.Set("ozdhacxtb@9", "xhqmcdqu") // <nil> #2958
        db.Get("nhxpejjq@1") // [""] pebble: not found #2959
        db.Set("eumlxphu@9", "") // <nil> #2960
        db.Merge("ystayojd@4", "gfeuscwqgld") // <nil> #2961
        db.SingleDelete("bmfprdgyll@8", false /* maybeReplaceDelete */) // <nil> #2962
        // INFO: [JOB 77] sstable deleted 000043
        // INFO: [JOB 77] sstable deleted 000046
        // INFO: [JOB 77] sstable deleted 000047
        // INFO: [JOB 77] sstable deleted 000050
        // INFO: [JOB 83] compacting(default) L0 [000054 000052] (2.4 K) + L6 [000051] (4.6 K)
        --- FAIL: TestMeta/execution/standard-013 (1.25s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1674719025301291349 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/db.go
+++ b/db.go
@@ -982,7 +982,7 @@ func (d *DB) newIterInternal(batch *Batch, s *Snapshot, o *IterOptions) *Iterato
 	}
 	if o != nil {
 		dbi.opts = *o
-		dbi.saveBounds(o.LowerBound, o.UpperBound)
+		dbi.processBounds(o.LowerBound, o.UpperBound)
 	}
 	dbi.opts.logger = d.opts.Logger
 	if d.opts.private.disableLazyCombinedIteration {

--- a/external_iterator.go
+++ b/external_iterator.go
@@ -143,7 +143,7 @@ func NewExternalIter(
 	}
 	if iterOpts != nil {
 		dbi.opts = *iterOpts
-		dbi.saveBounds(iterOpts.LowerBound, iterOpts.UpperBound)
+		dbi.processBounds(iterOpts.LowerBound, iterOpts.UpperBound)
 	}
 	for i := range extraOpts {
 		extraOpts[i].iterApply(dbi)

--- a/iterator.go
+++ b/iterator.go
@@ -307,6 +307,9 @@ type Iterator struct {
 	closePointIterOnce bool
 	// Used in some tests to disable the random disabling of seek optimizations.
 	forceEnableSeekOpt bool
+	// Set to true if NextPrefix is not currently permitted. Defaults to false
+	// in case an iterator never had any bounds.
+	nextPrefixNotPermittedByUpperBound bool
 }
 
 // cmp is a convenience shorthand for the i.comparer.Compare function.
@@ -1603,7 +1606,19 @@ func (i *Iterator) NextWithLimit(limit []byte) IterValidityState {
 //
 // It is not permitted to invoke NextPrefix while at a IterAtLimit position.
 // When called in this condition, NextPrefix has non-deterministic behavior.
+//
+// It is not permitted to invoke NextPrefix when the Iterator has an
+// upper-bound that is a versioned MVCC key (see the comment for
+// Comparer.Split). It returns an error in this case.
 func (i *Iterator) NextPrefix() bool {
+	if i.nextPrefixNotPermittedByUpperBound {
+		i.lastPositioningOp = unknownLastPositionOp
+		i.requiresReposition = false
+		i.err = errors.Errorf("NextPrefix not permitted with upper bound %s",
+			i.comparer.FormatKey(i.opts.UpperBound))
+		i.iterValidityState = IterExhausted
+		return false
+	}
 	if i.hasPrefix {
 		i.iterValidityState = IterExhausted
 		return false
@@ -2278,7 +2293,7 @@ func (i *Iterator) SetBounds(lower, upper []byte) {
 
 	// Copy the user-provided bounds into an Iterator-owned buffer, and set them
 	// on i.opts.{Lower,Upper}Bound.
-	i.saveBounds(lower, upper)
+	i.processBounds(lower, upper)
 
 	i.iter.SetBounds(i.opts.LowerBound, i.opts.UpperBound)
 	// If the iterator has an open point iterator that's not currently being
@@ -2300,7 +2315,10 @@ func (i *Iterator) SetBounds(lower, upper []byte) {
 	i.invalidate()
 }
 
-func (i *Iterator) saveBounds(lower, upper []byte) {
+// Initialization and changing of the bounds must call processBounds.
+// processBounds saves the bounds and computes derived state from those
+// bounds.
+func (i *Iterator) processBounds(lower, upper []byte) {
 	// Copy the user-provided bounds into an Iterator-owned buffer. We can't
 	// overwrite the current bounds, because some internal iterators compare old
 	// and new bounds for optimizations.
@@ -2312,9 +2330,19 @@ func (i *Iterator) saveBounds(lower, upper []byte) {
 	} else {
 		i.opts.LowerBound = nil
 	}
+	i.nextPrefixNotPermittedByUpperBound = false
 	if upper != nil {
 		buf = append(buf, upper...)
 		i.opts.UpperBound = buf[len(buf)-len(upper):]
+		if i.comparer.Split != nil {
+			if i.comparer.Split(i.opts.UpperBound) != len(i.opts.UpperBound) {
+				// Setting an upper bound that is a versioned MVCC key. This means
+				// that a key can have some MVCC versions before the upper bound and
+				// some after. This causes significant complications for NextPrefix,
+				// so we bar the user of NextPrefix.
+				i.nextPrefixNotPermittedByUpperBound = true
+			}
+		}
 	} else {
 		i.opts.UpperBound = nil
 	}
@@ -2499,7 +2527,7 @@ func (i *Iterator) SetOptions(o *IterOptions) {
 		i.opts.LowerBound, i.opts.UpperBound = lower, upper
 	} else {
 		i.opts = *o
-		i.saveBounds(o.LowerBound, o.UpperBound)
+		i.processBounds(o.LowerBound, o.UpperBound)
 		// Propagate the changed bounds to the existing point iterator.
 		// NB: We propagate i.opts.{Lower,Upper}Bound, not o.{Lower,Upper}Bound
 		// because i.opts now point to buffers owned by Pebble.
@@ -2630,7 +2658,7 @@ func (i *Iterator) Clone(opts CloneOptions) (*Iterator, error) {
 		newIterRangeKey:     i.newIterRangeKey,
 		seqNum:              i.seqNum,
 	}
-	dbi.saveBounds(dbi.opts.LowerBound, dbi.opts.UpperBound)
+	dbi.processBounds(dbi.opts.LowerBound, dbi.opts.UpperBound)
 
 	// If the caller requested the clone have a current view of the indexed
 	// batch, set the clone's batch sequence number appropriately.

--- a/testdata/iter_histories/next_prefix
+++ b/testdata/iter_histories/next_prefix
@@ -234,3 +234,59 @@ next-prefix
 a@10: (a@10, .)
 a@100: (a@100, .)
 b@100: (b@100, .)
+
+reset
+----
+
+populate keylen=1 timestamps=(1, 10, 100)
+----
+wrote 78 keys
+
+flush
+----
+
+lsm
+----
+0.0:
+  000005:[a@100#3,SET-z@1#76,SET]
+
+# Test for https://github.com/cockroachdb/pebble/issues/2260. Triggered the
+# bug. The second call to first would return c@100 instead of the correct key,
+# b@1.
+combined-iter upper=b@1
+first
+next-prefix
+next-prefix
+set-bounds lower=b@1 upper=d
+first
+next
+first
+----
+a@100: (a@100, .)
+err=NextPrefix not permitted with upper bound b@1
+err=NextPrefix not permitted with upper bound b@1
+.
+b@1: (b@1, .)
+c@100: (c@100, .)
+b@1: (b@1, .)
+
+# Did not trigger https://github.com/cockroachdb/pebble/issues/2260 since
+# Iterator.NextPrefix first does a Next. So the second call to NextPrefix
+# returned after the Next, since the upper bound was reached, which left the
+# Iterator positioned at b@1.
+combined-iter upper=b@10
+first
+next-prefix
+next-prefix
+set-bounds lower=b@1 upper=d
+first
+next
+first
+----
+a@100: (a@100, .)
+err=NextPrefix not permitted with upper bound b@10
+err=NextPrefix not permitted with upper bound b@10
+.
+b@1: (b@1, .)
+c@100: (c@100, .)
+b@1: (b@1, .)

